cfgmgr/vlanmgr.cpp

```diff
@@ -0,0 +1,488 @@
+    // When port is not member of any VLAN, it shall be detached from Dot1Q bridge!
+    cmd << BRIDGE_CMD << " vlan show dev " << port_alias << " | " << GREP_CMD << " None";
+    EXEC_WITH_ERROR_THROW(cmd.str(), res);
+    if (!res.empty())
+    {
+        cmd.str("");
+        cmd << IP_CMD << " link set " << port_alias << " nomaster";
+        EXEC_WITH_ERROR_THROW(cmd.str(), res);
+    }
+
+    return true;
+}
+
+bool VlanMgr::isVlanMacOk()
+{
+    return !(!gMacAddress);
+}
+
+void VlanMgr::doVlanTask(Consumer &consumer)
+{
+
+    if (!isVlanMacOk())
+    {
+        SWSS_LOG_DEBUG("VLAN mac not ready, delaying VLAN task");
+        return;
+    }
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        auto &t = it->second;
+
+        string key = kfvKey(t);
+
+        /* Ensure the key starts with "Vlan" otherwise ignore */
+        if (strncmp(key.c_str(), VLAN_PREFIX, 4))
+        {
+            SWSS_LOG_ERROR("Invalid key format. No 'Vlan' prefix: %s", key.c_str());
+            it = consumer.m_toSync.erase(it);
+            continue;
+        }
+
+        int vlan_id;
+        vlan_id = stoi(key.substr(4));
+
+        string vlan_alias, port_alias;
+        string op = kfvOp(t);
+
+        if (op == SET_COMMAND)
+        {
+            string admin_status;
+            uint32_t mtu = 0;
+            vector<FieldValueTuple> fvVector;
+            string members;
+
+            /* Add host VLAN when it has not been created. */
+            if (m_vlans.find(key) == m_vlans.end())
+            {
+                addHostVlan(vlan_id);
+            }
+
+            /* set up host env .... */
+            for (auto i : kfvFieldsValues(t))
+            {
+                /* Set vlan admin status */
+                if (fvField(i) == "admin_status")
+                {
+                    admin_status = fvValue(i);
+                    setHostVlanAdminState(vlan_id, admin_status);
+                    fvVector.push_back(i);
+                }
+                /* Set vlan mtu */
+                else if (fvField(i) == "mtu")
+                {
+                    mtu = (uint32_t)stoul(fvValue(i));
+                    /*
+                     * TODO: support host VLAN mtu setting.
+                     * Host VLAN mtu should be set only after member configured
+                     * and VLAN state is not UNKNOWN.
+                     */
+                    SWSS_LOG_DEBUG("%s mtu %u: Host VLAN mtu setting to be supported.", key.c_str(), mtu);
+                    fvVector.push_back(i);
+                }
+                else if (fvField(i) == "members@") {
+                    members = fvValue(i);
+                }
+            }
+            /* fvVector should not be empty */
+            if (fvVector.empty())
+            {
+                FieldValueTuple a("admin_status",  "up");
+                fvVector.push_back(a);
+            }
+            m_appVlanTableProducer.set(key, fvVector);
```

Comment:
Is there any plan to deal with the operational state of Vlan interface if not yet done?  We may need that to disable routing if vlan interface is down.
If we are doing that, I assume some other entity (like netlink message in linksync.cpp) will update the app-db?  In that case, do we have a timing issue here? (e,g, netlink messages came before we put the table to app-db)?

We can address this in future PR if needed.

Reply:
we need to define the operational state of vlan first. if it mean all members are down, then we have way to know the member oper status already, we do not need to rely on the netlink to get the vlan op status, then. This is going to be another feature request and thus a separate PR.